File: Java/src/MergeIntervals.java
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;

/**
 * Given a collection of intervals, merge all overlapping intervals.
 * <p>
 * For example,
 * Given [1,3],[2,6],[8,10],[15,18],
 * return [1,6],[8,10],[15,18].
 * <p>
 * Accepted.
 */
public class MergeIntervals {

    public List<Interval> merge(List<Interval> intervals) {
        if (intervals == null || intervals.isEmpty() || intervals.size() == 1) {
            return intervals;
        }

        List<Interval> results = new LinkedList<>();

        // Sort the list by starts.
        intervals.sort(Comparator.comparingInt(o -> o.start));

        results.add(intervals.get(0));
        for (int i = 1; i < intervals.size(); i++) {
            int resultsSize = results.size();
            Interval lastResult = results.get(resultsSize - 1);
            if (lastResult.end >= intervals.get(i).start) {
                results.get(resultsSize - 1).end = Math.max(lastResult.end, intervals.get(i).end);
            } else {
                results.add(intervals.get(i));
            }
        }

        return results;
    }

    public static class Interval {

        int start;
        int end;

        Interval() {
            start = 0;
            end = 0;
        }

        Interval(int s, int e) {
            start = s;
            end = e;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof Interval) {
                Interval interval = (Interval) obj;
                return this.start == interval.start && this.end == interval.end;
            }
            return false;
        }

    }

}

File: JavaScript/src/MergeIntervals.js
/**
 * Given a collection of intervals, merge all overlapping intervals.
 *
 * For example,
 * Given [1,3],[2,6],[8,10],[15,18],
 * return [1,6],[8,10],[15,18].
 *
 * Accepted.
 */

function Interval(start, end) {
    this.start = start;
    this.end = end;
}


/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */
let merge = function (intervals) {
    if (intervals == null || intervals.length === 0 || intervals.length === 1) {
        return intervals;
    }

    let results = [];

    // Sort the list by starts.
    intervals.sort(function (a, b) {
        if (a.start > b.start) {
            return 1;
        } else if (a.start === b.start) {
            return 0;
        }
        return -1;
    });

    results.push(intervals[0]);
    for (let i = 1; i < intervals.length; i++) {
        let resultsSize = results.length;
        let lastResult = results[resultsSize - 1];
        if (lastResult.end >= intervals[i].start) {
            results[resultsSize - 1].end = Math.max(lastResult.end, intervals[i].end);
        } else {
            results.push(intervals[i]);
        }
    }

    return results;
};


if (merge([new Interval(1, 1)]).toString() === [new Interval(1, 1)].toString()) {
    console.log("pass")
} else {
    console.error("failed")
}

if (merge([new Interval(1, 2), new Interval(2, 2)]).toString() === [new Interval(1, 2)].toString()) {
    console.log("pass")
} else {
    console.error("failed")
}

if (merge([new Interval(1, 2), new Interval(4, 5), new Interval(7, 8)]).toString() === [new Interval(1, 2), new Interval(4, 5), new Interval(7, 8)].toString()) {
    console.log("pass")
} else {
    console.error("failed")
}

if (merge([new Interval(1, 3), new Interval(2, 4)]).toString() === [new Interval(1, 4)].toString()) {
    console.log("pass")
} else {
    console.error("failed")
}

if (merge([new Interval(1, 4), new Interval(2, 3)]).toString() === [new Interval(1, 4)].toString()) {
    console.log("pass")
} else {
    console.error("failed")
}